bsip-0044.md

```diff
@@ -0,0 +1,280 @@
+
+Bob queries the BitShares Network for the `contract identifier` Alice provided. He examines to ensure it contains his desired `recipient` account, `asset` symbol, asset `quantity`, `preimage length`, and `timelock threshold`. Bob now creates and funds his own HTLC on the Bitcoin Network that will spend the `UTXO` of this contract to the `recipient address` Alice provided during their setup phase, of `amount` 1 BTC if a `preimage` that generates the `preimage hash` Bob copied from Alice's HTLC before the `timelock threshold` of 5pm today. Upon consensus validation of Bob's HTLC on the Bitcoin Network, 1 BTC he controlled are spent into the contract and "locked". He then shares the resulting `contract identifier` with Alice. Notice Bob specified a `timelock threshold` much shorter than Alice defined in her contract. This ensures Bob will have enough time to observe and use the `preimage` Alice will publish to the blockchain next. 
+
+Alice now examines the HTLC Bob created on the Bitcoin Network, ensuring the `preimage hash` and `preimage length` both match the original values she used within her contract. She also verifies her desired `recipient address`, `quantity`, and `timelock threshold` agree with her intentions. She now uses her `preimage` to "unlock" Bob's contract. Once consensus validation occurs on the Bitcoin Network, the HTLC will spend 1 BTC to Alice's `recipient address`. This reveals the `preimage` on the Bitcoin Network for Bob to use next. NOTE: She must do this before 5PM. Otherwise, Bob may (and should) reclaim the funds in the contract he created.
+
+Bob has now observed the `preimage` Alice used to "unlock" his HTLC, and he will use it to "unlock" her HTLC to receive the 10,000 bitUSD `secured assets` into his `recipient` account "bob". NOTE: He must do this before 10AM tomorrow. Otherwise, Alice may (and should) reclaim the funds in the contract she created.
+
+# **Specifications**
+
+## **Objects**
+
+```
+htlc_obj
+  id                // int
+  depositor         // account object
+  quantity          // int
+  symbol            // string
+  recipient         // account object
+  hash_algorithm    // enum
+  preimage_hash     // string
+  preimage_length   // int
+  preimage_tx_id    // reference to transaction object containing the preimage used for htlc_redeem
+  timeout_threshold // datetime
+
+  // Constructor
+  Set: id = next available
+  Initialize other fields = NULL
+
+  return contract
+```
```

Comment:
Thanks for the proper code. I can add this in. 

As for the reference rather than the preimage data within the HTLC object: my desire is to store this (rather larger) data object only once in the blockchain. I’m not a real developer, so will defer to you guys on implementation. My thought is: the Redeem TX must contain the valid preimage to be validated, so that is when and only when it needs to be committed to the blockchain. Can have the operation update the object to reference the TX. Can that work?

Reply:
I agree that the crux of the issue is storage on the blockchain. The potential size of the preimage means this should be carefully considered. The operation that creates the HTLC tells us the preimage size, but not the preimage. A later transaction (that may never happen) would contain the actual preimage.

The immutability of the chain means that a redeem operation cannot go back and "fill in" the preimage there. Hence, the redeem operation must store the preimage on its own. But I believe with account history, an htlc object should be modified to include the preimage. That allows for clients to request 1 object that has the complete picture. If it only included the transaction id of the preimage, a second call would be necessary.

Please shoot holes in my theory.

Reply:
Thinking it over again with the help of @pmconrad I can see an advantage to only including a reference to the transaction id. Yes, it could potentially cause the client to do 2 lookups, but that also means that the preimage does not have to be stored in the live database, as it is on the chain.

Once again, please weigh in with your thoughts.

Reply:
I'm going to offer my thoughts, with the caveat these concepts are beyond my authoritative knowing, so happy to be corrected: It is my understanding that the `recipient` will likely "subscribe to the HTLC" created by `depositor` after they evaluate the contents match their desires. This will allow the `recipient` to receive notification once the `depositor` supplies the `preimage` in the Redeem operation. I don't know enough about which bits get stored where to shoot holes.

Reply:
In the few other HTLC implementations I have read about, the depositor shares the preimage out-of-band. That gives the redeemer the option to execute the contract or not. If he decides to not "sign on the dotted line", he simply does not use the preimage, and the contract times out.

Reply:
This is my understanding of HTLC process flow:

Given: d depositor, r receiver, p preimage, h hash(p), l length(p), c HTLC, A Alice, B Bob
A: generate p, h, l
A: create A_c as A_d >>> B_r, h, length(p)
A: send A_c identifier to B
B: evaluate A_c
B: create B_c as B_d >>> A_r, h, length(p)
A: receive B_c identifier
A: evaluate B_c
A: redeem B_c with p
B: observe p and redeem A_c with p

A will reveal p only when redeem B_c. If B becomes aware of p, B may allow B_c to expire and redeem A_c prior to expiry without penalty.